### ISSUE TITLE
ttnn.add/subtract/multiply silently return wrong values at rank >= 7 when an operand is broadcast at dim -6: binary_ng collapses dims -6 and outward into one counter, and four of six tiles read past the end of the operand

### ISSUE THREAD

**Issue by singhharsh1708** [tenstorrent/tt-metal#55351]
**Component / Area:** ttnn eltwise binary (binary_ng)

**Issue Type:** Bad Outputs

## Summary

At rank >= 7, `ttnn.add` / `ttnn.subtract` / `ttnn.multiply` return silently wrong values whenever an operand is broadcast along dimension -6 — a broadcast the op's own validation explicitly permits. Five of six output tiles are wrong in the smallest reproducing case, and four of them come from reading past the end of the operand's buffer.

Measured on ttsim v1.10.3 (`libttsim_bh.so`, Blackhole), tt-metal `v0.79.0-dev20260902` debs, bfloat16, TILE layout, DRAM interleaved. Every value below is a power of two or a small multiple of ten, exact in bfloat16, so nothing here is a rounding artefact.

```python
a = torch.zeros((2,1,1,1,1,32,32)); a[0] = 1.0;  a[1] = 2.0
b = torch.zeros((2,3,1,1,1,32,32)); b[:,0] = 10.0; b[:,1] = 20.0; b[:,2] = 40.0
ttnn.add(a, b)
```

| | ttnn | torch |
|---|---|---|
| `out[0]` | `[11, 22, 40]` | `[11, 21, 41]` |
| `out[1]` | `[10, 20, 40]` | `[12, 22, 42]` |

The output shape is correct; only the values are wrong. It happens in both operand positions — with the broadcast on the second operand instead, `a(2,3,…) + b(2,1,…)` gives `[[11, 22, 80], [20, 40, 80]]` against torch's `[[11, 21, 41], [12, 22, 42]]`, again 5 of 6 tiles.

The same shapes one rank lower are correct:

| case | result |
|---|---|
| rank 6, `a(2,1,1,1,32,32) + b(2,3,1,1,32,32)` | correct |
| rank 6, `a(1,1,1,1,32,32) + b(3,3,1,1,32,32)` | correct |
| rank 7, no broadcast | correct |
| **rank 7, broadcast at dim -6** | **wrong** |
| rank 7, broadcast at dim -7 | correctly rejected with `TT_FATAL` |

## Cause

`extract_nD_dims` ([`binary_ng_program_factory.cpp#L30-L40`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp#L30-L40)) multiplies every dimension from -6 outward into one scalar:

```cpp
// For rank > 5 dims will be collapsed into a single dim
uint32_t extract_nD_dims(const Tensor& x, const int out_rank) {
    const auto& shape = x.logical_shape();
    uint32_t nD_dim = 1;
    if (out_rank >= 6 && shape.rank() >= 6) {
        for (int i = -6; i >= -out_rank; --i) {
            auto dim = shape[i];
            nD_dim *= dim;
        }
    }
    return nD_dim;
}
```

The reader's stride for that whole group is then chosen by a single boolean ([#L797](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp#L797) and [#L808](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_program_factory.cpp#L808)):

```cpp
aHt * aWt * aC * aN * aD * (aND > 1),
...
bHt * bWt * bC * bN * bD * (bND > 1),
```

That boolean has exactly two states: broadcast the entire collapsed group, or stride through all of it. It cannot express "broadcast at -6 but stride at -7", which is precisely what a rank-7 tensor with a unit dim at -6 requires.

In the example above, `a` has `shape[-6] = 1` and `shape[-7] = 2`, so `aND = 1 * 2 = 2`. Because `aND > 1`, the guard emits a full stride, and the reader walks `a`'s tiles as `0, 1, 2, 3, 4, 5` for the six output tiles where it should walk `0, 0, 0, 1, 1, 1`. `a` holds only two tiles (`2 * 32 * 32 = 2048` elements), so four of those six fetches are past the end of the buffer — which is where the zeros in `out[1]` come from.

The measured values match that walk exactly: `out[0][1] = 22 = a[1] + b[1]` (should be `a[0] + b[1] = 21`), and everything from the third tile on reads out of bounds.

## This shape is supposed to work

The device op's validation ([`binary_ng_device_operation.cpp#L445-L453`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp#L445-L453)) rejects broadcasting only at index -7 and beyond:

```cpp
if (i <= -7) {
    TT_FATAL(
        a_dim == b_dim,
        "Broadcasting rule violation for rank >= 7 at dimension index {} (output rank {}). "
        "Broadcast is supported up to rank 6. dim a: {}, dim b: {}", ...);
}
```

So dimension -6 is deliberately inside the supported set — "Broadcast is supported up to rank 6" — and the guard lets this shape through on purpose. Broadcast at -6 was added intentionally in #41272 ("Extend broadcast support from rank 5 to rank 6 for binary and ternary"). It works at rank 6, where the collapsed group contains only dim -6. It breaks at rank 7 and above, where the same flat counter has to carry dim -7 as well.

There is no fallback: nothing re-reads or repairs the operand afterwards, and the wrong tiles are written straight to the output.

## Why tests miss it

The rank-6 case works, and that is where the broadcast tests for this feature live. Exercising the defect needs rank >= 7 *and* a unit dim at -6 *and* a non-unit dim further out — the combination that makes the collapsed extent greater than one while still requiring a broadcast inside it. Any rank-7 test with no broadcast passes, and any test that broadcasts at -7 hits the `TT_FATAL` and is presumably written as a negative test.

## Fix

The immediate, zero-cost fix is to make the validation match what the program factory can actually do — extend the existing guard so that a broadcast anywhere inside the collapsed group is rejected when the group spans more than one dimension. That turns a silent wrong answer into the same clear `TT_FATAL` the -7 case already produces, and is a change to one condition in a file already in scope.

The complete fix is to give the nD group a real per-dimension stride rather than one collapsed counter and one boolean, so that broadcast at -6 keeps working at any rank, as the validation message promises.

Either is fine by me; the important part is that `ttnn.add` should not read past the end of an operand and return the result.

## Validation

All numbers above are from ttsim v1.10.3 running the real ops through the published debs; no hardware. Confirmed on `add`, `subtract` and `multiply`, in both operand positions, with rank-6 and no-broadcast rank-7 controls, and with all operand values chosen to be bfloat16-exact.

I would like to work on this through the bounty program, and am happy to prepare the PR.
